src/System.Management.Automation/engine/CommandProcessorBase.cs

```diff
@@ -644,13 +644,86 @@ internal void DoComplete()
                 _context.ShellFunctionErrorOutputPipe = oldErrorOutputPipe;
                 _context.CurrentCommandProcessor = oldCurrentCommandProcessor;
 
+                // Restore the previous scope...
+                if (_previousScope != null)
+                {
+                    // Restore the scope but use the same session state instance we
+                    // got it from because the command may have changed the execution context
+                    // session state...
+                    CommandSessionState.CurrentScope = _previousScope;
+                }
+
+                // Restore the previous session state
+                if (_previousCommandSessionState != null)
+                {
+                    Context.EngineSessionState = _previousCommandSessionState;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Executes the appropriate disposal methods for script commands.
+        /// <see cref="DlrScriptCommandProcessor"/> is itself a command processor, so interpreted functions cannot be
+        /// reliably disposed without also disposing the entire command processor. Instead, it implements its own
+        /// <see cref="DlrScriptCommandProcessor.InvokeCleanupBlock"/> method to perform the same function as its
+        /// compiled counterpart <see cref="PSScriptCmdlet.Dispose"/>, which is a proper IDisposable implementation.
+        /// </summary>
+        internal virtual void CleanupScriptCommand()
+        {
+            var scriptCmdlet = Command as PSScriptCmdlet;
+            var scriptProcessor = this as DlrScriptCommandProcessor;
+
+            if (scriptCmdlet == null && scriptProcessor == null)
+            {
+                return;
+            }
```

Comment:
This is already in the base class (`CommandProcessorBase`). I'm not entirely sure what you mean by that 😅.

If you mean, can I move it to `DlrScriptCommandProcessor` or `ScriptCommandProcessor` I believe that's not possible, as those aren't used for script cmdlets, only script functions as far as I understand that section of code.

`PSScriptCmdlet` is a `PSCmdlet`, which... I don't _think_ inherits from `CommandProcessorBase` itself? Whereas `DlrScriptCommandProcessor` does inherit from `CommandProcessorBase`. So short of completely duplicating the code and having the command processor handle it for script functions but the generated cmdlet handle it for script cmdlets, I don't see a better way to do this one.

I can't really speak to whether making `CommandProcessorBase` a generic / fully abstract class is the right move here, I think it's _possible_ that would improve this a fair bit (as well as potentially other areas of the code), but if we want to do that it would need to be a wholly separate PR with some fairly extensive review to ensure we don't break anything. I don't think we should do that in a first pass here, it seems too liable to break a lot more things.

Reply:
That said, I'm happy to poke at making that a reality. I don't know quite how extensive that change would be without having a proper attempt at it first, I'm sure. 😂

Reply:
All I mean here is:

- There's a `virtual` method (implying the logic can be overridden by a subclass)
- In that method, we test if we *are* a subclass and do nothing otherwise

That second thing is exactly what virtual methods are for, but with a much more efficient implementation:

```csharp
class CommandProcessorBase
{
    internal virtual void CleanupScriptCommand()
    {
        // Do nothing -- subclasses may use
    }
}

class DlrScriptCommandProcessor : CommandProcessorBase
{
    internal override void CleanupScriptCommand()
    {
        // Actual logic for scriptCmdlets
    }
}
```

Reply:
Hmm, fair. We could opt for something more like this then?

```csharp
class CommandProcessorBase
{
    protected virtual void CleanupScriptCommand()
    {
        if (this.Command is PSScriptCmdlet) {
        {
            // cleanup script cmdlet
        }
    }
}

class DlrScriptCommandProcessor : CommandProcessorBase
{
    protected override void CleanupScriptCommand
    {
        // cleanup function
    }
}
```

Reply:
🤔 tho I guess if I could have the base method be empty and call the PSScriptCmdlet disposal in the Dispose() method proper if I shift some of the logic about a bit.